### PR TITLE
More structural comparisons during variance computation

### LINKED ISSUE #44572: Check order dependence with mutually-recursive non-unary generics
# Bug Report

Adding an implementation of an interface allows an invalid assignability check of the interface to pass

### 🔎 Search Terms

assignability, unnecessary implementation

### 🕗 Version & Regression Information

- This changed between versions 3.7.5 and 3.8.3

It's still present in all versions including nightly.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.3.2&strict=true#code/JYOwLgpgTgZghgYwgAgApyhcAeAggGmQCEA+ZAbwChlkBzCMAYQAtgAbAEwAoBKALmQt2HPIVIBuStWTBIUXgICScuACM2EbOkw4CxEiUkBfKaDnwkg1p1H7pEAB6QQHAM5oMWMLdIVpmOA4AexA2AE9kOAFcSRoA4NCI1QEiYykENjhXd0UAWwAHNh8yYAKNXK93bS9ivxoEENcwKABXBDAg+XiQ8OQEaw4BIRs9Uh465BNpeiYBhSthWqoaOIYWqBBkMFZXADp+4VjJqRoAKlloeeVoNQ0tT10xAwn6xrBkXLh85ABeZBAIAB3ZAAWS+2GGIhaIAA1iAgoCQIRoXCESASIRIcUANoAXRIvEk0hoMGh7WAIVOdCwXGJK1A+SGA2KdP4yGuUFumixo2eyxWNDCwAgnCpn3yuxmXAZ4wAPrLkHijjQTCrKFMGiAmsgHAJqjgUfDEcjYUb0b8KEZIu44CAwpJNdqAPp6h7eEAtNhsE2oxFkP4OcRAA)

Thanks @MartinJohns for the link

### 💻 Code

```ts
interface Parent<A, B> {
  getChild(): Child<A, B>;

  iter(): Iterable<Parent<A, B>>;
}

interface Child<A, B>
  extends Parent<A, B> {
  readonly a: A;
  readonly b: B;
}

class Impl<A, B> implements Parent<A, B> {
  constructor(readonly child: Child<A, B>) {
  }

  getChild(): Child<A, B> {
    return this.child;
  }

  *iter(): Iterable<Parent<A, B>> {
    const map = new Map<Child<unknown, unknown>, Child<A, B>[]>();

    function* gen(
      inp: Child<A, B>
    ): Iterable<Child<A, B>> {
      yield* map.get(inp) || [];
    }
  }
}

const x: Parent<unknown, unknown> = {} as any;
const _: Parent<null, unknown> = x; // should not pass
```

The final assignment should not pass. It accurately errors in 3.7.5,  if the `Impl` class is removed (or any aspect of the `iter` implementation is modified), or by adding a sentinel type like `sentinel?: A` to the `Parent` interface to aid in type checking.

### 🙁 Actual behavior

No error is thrown in the final assignment.

### 🙂 Expected behavior

A n error is throw:
```
Type 'Parent<unknown, unknown>' is not assignable to type 'Parent<null, unknown>'.
  Type 'unknown' is not assignable to type 'null'.
```

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR fixes variance computation to perform structural comparisons for all type alias instantiations and object type instantiations with one or more variance markers as type arguments. Further discussion of the fix [here](https://github.com/microsoft/TypeScript/pull/45578#issuecomment-907888926).

Fixes #44572.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot perf test this

**Comment by ahejlsberg** [comment]
Performance looks good, even slightly improved, likely because we're no longer jamming an `aliasTypeArgumentsContainsMarker` property onto variance measurement type instantiations.

**Comment by ahejlsberg** [comment]
The failed fourslash test appears to be a preexisting problem in the main branch.

File: src/compiler/checker.ts
Comment (andrewbranch) [inline src/compiler/checker.ts:12915]: Why not just add `MarkerType` to `PropagatingFlags`?

**Review by andrewbranch — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt

File: src/compiler/checker.ts
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:12905]: This check happens a few times - maybe just hoist it out to its own function and call it in `reportUnmeasurableMarkers`/`reportUnreliableMarkers`.

File: src/compiler/checker.ts
Comment (weswigham) [reply src/compiler/checker.ts:12915]: I agree - it was a flag on the type/list before so we could avoid doing any kind of deep checks for markers and catch pretty much the exact type we initially make. It was a known shortcoming that we missed markers that flowed into other versions of the reference we're checking the variance of (known to me, at least - I thought it was an intentional performance tradeoff). This now catches if a `Whatever<?, T>` has a `Whatever<T, ?>` dependency (and forces structural comparison, rather than assuming covariance), but still misses a `Whatever<{x: ?}, T>` dependency. Shouldn't we go all the way and catch if the marker appears in the type argument list in any way?

Comment (ahejlsberg) [reply src/compiler/checker.ts:12915]: Ideally we'd identify any occurrence of a marker type, but it would require eager exploration which we know doesn't end well for many classes of types. Our variance measurement is effectively a balance between correctness and performance, and here we're moving a little further towards correctness. I know we've tried doing _only_ structural comparisons when computing variance, but it isn't feasible. It just reveals the same problems with (almost) infinite types that we're trying to solve by doing variance measurement in the first place.

Comment (ahejlsberg) [reply src/compiler/checker.ts:12915]: We could add `MarkerType` to `PropagatingFlags`, but it wouldn't save us much and would require us to also include `TypeFlags.TypeParameter` in `TypeFlags.ObjectFlagsType`, which would mean an extra property on every instance that we otherwise don't need.

**Comment by Andarist** [comment]
Would it be possible to generate a playground for this PR?

**Comment by andrewbranch** [comment]
I forget whether it checks out the head or merge commit—if the former, it would need its merge conflicts fixed, but we can try.

@typescript-bot pack this

**Comment by andrewbranch** [comment]
@ahejlsberg @DanielRosenwasser did we have any big concerns over this or did we mostly forget about it? If we think we might want to merge it, let’s work on getting it in early for 4.7.

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster

**Comment by Andarist** [comment]
@andrewbranch could you request a new tarball? the spawned task has failed but that was executed before the recent merge, perhaps a fresh build after this merge would succeed

**Comment by ahejlsberg** [comment]
@typescript-bot pack this

**Comment by ahejlsberg** [comment]
Yikes, the `xstate` performance test is heavily impacted by this (+200%). Definitely not ready to go yet.

**Comment by Andarist** [comment]
Hah, I'm actually XState maintainer 😅 In a way - I think that our types need a solid overhaul *anyway*. I wonder if there is anything in particular that we could take into consideration when refactoring, are there any particular things that we could try to avoid. While... we are flattered to be included in your perf test suite 😉 it's also a good sign that maybe we are overcomplicating things for the typechecker. 

Some recent changes to our types are in particular quite crazy (those designed to work with the typegen and have very unusual patterns used in them to "stitch" types together). My sincere hope is that this particular perf problem is not related to this part of the types as I took a loooot of time to get them working. Could you tell me against which version of XState are you running this benchmark? It would be useful for such information to be included somewhere in the result - as far as I can tell it's not in the comment of the bot nor in the attached benchmark file.

When it comes to the variance itself - I'm pretty sure our types "get it wrong" (from the user perspective). While I understand the concept of variance and can "compute" variance type in simple cases, I have no idea how I should think about it with such complicated types as ours, with a lot of layers, a lot of generics etc. Are there any resources/comments from the past that could help me understand this? Are there any tricks that could help TS to determine the variance type "quicker"?

On a more positive note - I've tested this build and it seems to fix my minimal repro case reported [here](https://github.com/microsoft/TypeScript/issues/44572#issuecomment-1042915921) and it also fixes the issue in our repo from which I've distilled this minimal repro. 🎉

**Comment by RyanCavanaugh** [comment]
@Andarist Thanks for taking such an interest in this!

Our current instructions for profiling compilation times are at https://github.com/microsoft/TypeScript/wiki/Performance-Tracing and there are some other resources at https://github.com/microsoft/TypeScript/wiki/Performance

The benchmarks run against whatever is currently checked into GitHub. We run against latest to avoid getting stale versions of projects which might not be taking advantage of newer language features.

How to make a project build faster in general is outside the scope of this comment box, but I'd be happy to talk about it in a separate issue after you've run tracing and identified some hotspots.

**Comment by Andarist** [comment]
@RyanCavanaugh would you recommend performing this profiling using the latest TS or using the build from this PR? I'm unsure what's better from your perspective - for us to attempt to recognize current issues~ or the stuff specifically related to this PR?

> The benchmarks run against whatever is currently checked into GitHub. We run against latest to avoid getting stale versions of projects which might not be taking advantage of newer language features.

Thanks, that's good to know - I would recommend adding information somewhere about the commit against which those profiling sessions were executed. It would ensure that we can always investigate exactly the same builds etc when any issues are recognized by this automatic benchmarking.

> How to make a project build faster in general is outside the scope of this comment box, but I'd be happy to talk about it in a separate issue after you've run tracing and identified some hotspots.

Thanks, I will certainly come back asking questions 🤣 I'll try to book some time in our next cycle to poke around this

**Comment by Andarist** [comment]
I might also have found a problem with this PR unless the goal of this PR is not to handle all of the possible cases. Are perhaps still situations when you might bail out from the process and assume that some types are assignable when they shouldn't?

As far as I can tell the result of the `spawn` should not be accepted within the `createMachine` call at the bottom of this screenshot. I've performed various checks above this situation to confirm subtyping rules etc

Screenshot 2022-02-24 at 11 26 55

I can try to prepare a reduced test case but since that's sometimes up to a day of work I'd like to confirm with you that it's worth it.

The issue can be observed on commit f256378e6 in xstate repo (this is a note for my own reference)

**Comment by ahejlsberg** [comment]
Closing in favor of #48080.

**Pull request closed without merging**
